### PR TITLE
update qcprot to cython.floating types

### PR DESCRIPTION
**Pull request by richardjgowers**
use memoryviews throughout for potential performance

allows either float32 or float64 inputs

Related #3927 

Changes made in this Pull Request:
 - qcprot functions take either float32 or float64 inputs
 - modernised qcprot to use memoryviews

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

## Developers certificate of origin
- [ ] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
:books: Documentation preview :books:: https://mdanalysis--4273.org.readthedocs.build/en/4273/

### CODE REVIEW

**Comment by richardjgowers** [comment]
Still need to do some tests here.  `rms.RMSD` looks like it's converting/copying everything up to float64, but I imagine that not forcing this would be some sort of precision-regression.  Instead I can probably write some tests that using `lib.qcprot` directly with floats/doubles work as expected and we can have some future discussion about precision and dtypes in the future.

**Review by IAlibay — APPROVED** [review]
This looks reasonable to me. Is there an associated issue this is working towards? (edit: I know there were a few out there that had been opened about fp precision).

**Comment by richardjgowers** [comment]
@IAlibay I've tagged a related issue, I think we should be moving towards numpy-like behaviour where output dtype matches input dtype and either float or double are allowed.  A future discussion is the next major release of MDA will probably have a single precision **option**, since we're able to get 2x performance & 1/2 memory footprint for this.

The memoryview change is just a best-practice and we should gradually modernise to this as it's documented as the recommended route for best performance.

File: package/CHANGELOG [hunk 1]
```diff
@@ -13,7 +13,7 @@ The rules for this file:
   * release numbers follow "Semantic Versioning" http://semver.org
 
 ------------------------------------------------------------------------------
-??/??/?? IAlibay, ianmkenney, PicoCentauri
+??/??/?? IAlibay, ianmkenney, PicoCentauri, richardjgowers
```
Comment (IAlibay) [inline package/CHANGELOG]: Odd, I thought you had conttributed in 2.6 for the GSD stuff, maybe that was 2.5... time flies by.

File: package/MDAnalysis/lib/qcprot.pyx [hunk 2]
```diff
@@ -494,7 +509,7 @@ def FastCalcRMSDAndRotation(np.ndarray[np.float64_t, ndim=1] rot,
                     rot[0] = rot[4] = rot[8] = 1.0
                     rot[1] = rot[2] = rot[3] = rot[5] = rot[6] = rot[7] = 0.0
 
-                    return
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/qcprot.pyx:497]: this branch looks like it was a bug.  at this point the RMSD has been calculated and the code has failed to determine a suitable rotation matrix, returning None doesn't seem to make sense

Comment (richardjgowers) [reply package/MDAnalysis/lib/qcprot.pyx:497]: e.g. compared to this implementation in BioPython: https://github.com/biopython/biopython/blob/master/Bio/PDB/qcprot.py#L190

File: package/MDAnalysis/lib/qcprot.pyx [hunk 3]
```diff
@@ -174,27 +176,25 @@ cdef extern from "math.h":
...
-def InnerProduct(np.ndarray[np.float64_t, ndim=1] A,
-                 np.ndarray[np.float64_t, ndim=2] coords1,
-                 np.ndarray[np.float64_t, ndim=2] coords2,
-                 int N,
-                 np.ndarray[np.float64_t, ndim=1] weight):
+cpdef cython.floating InnerProduct(cython.floating[:] A,
+                                   cython.floating[:, :] coords1,
+                                   cython.floating[:, :] coords2,
+                                   int N,
+                                   cython.floating[:] weight):
```
Comment (tylerjereddy) [inline package/MDAnalysis/lib/qcprot.pyx:183]: I remember that a while back I tried to improve performance in SciPy by switching from the old NumPy buffer syntax, to memoryviews, and it actually caused a performance regression: https://github.com/scipy/scipy/issues/12379#issuecomment-645934588

Anyway, you may want to measure that you remain at least performance neutral, my experience with following that best practice hasn't historically been great, but hopefully the situation is better now.

File: testsuite/MDAnalysisTests/lib/test_qcprot.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        r = qcprot.CalcRMSDRotationalMatrix(p1, p2, len(prot), rot, weights)
+
+        rot_ref = np.array([0.999998, 0.001696, 0.001004,
+                            -0.001698,  0.999998,  0.001373,
+                            -0.001002, -0.001375,  0.999999],
+                           dtype=dtype)
+
+        err = 0.001 if dtype is np.float32 else 0.000001
+        assert r == pytest.approx(0.6057544485785074, abs=err)
+        assert_array_almost_equal(rot, rot_ref)
```
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/lib/test_qcprot.py]: Maybe use `assert_allclose` for new test code per the `Note` at https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_almost_equal.html ?

**Review by orbeckst — CHANGES_REQUESTED** [review]
The RMSD functions are quite time-critical for many applications so if there's any chance for a performance impact then please provide some benchmark data before merge. Thanks!

**Comment by orbeckst** [comment]
About float vs double. I remember vaguely that during one of the REU projects we played around with single vs double precision and arrived at the conclusion that float32 is insufficient for the algorithm to properly converge. Of course, I don't remember any details and have no idea if there are any notes anywhere. Primarily I want to say that it simply might not be as easy to say that we just run the algorithm at whatever dtype the rest of the code wants to run and that this will require careful testing.

**Comment by richardjgowers** [comment]
@orbeckst benchmarks
```python
from MDAnalysisTests.datafiles import PSF, DCD
import MDAnalysis as mda
from MDAnalysis.lib import qcprot
import numpy as np

u = mda.Universe(PSF, DCD)

prot = u.select_atoms('protein')

w = (prot.masses / np.mean(prot.masses)).astype(np.float64)

ref = prot.positions.astype(np.float64)
u.trajectory[1]
conf = prot.positions.astype(np.float64)

rot = np.zeros(9, dtype=np.float64)
A = np.zeros(9)

N = len(prot)
```
then:
```python
%%timeit

qcprot.CalcRMSDRotationalMatrix(ref, conf, len(prot), rot, w)
```
old code:
`10.1 µs ± 46.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)`
new code:
`9.97 µs ± 29.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)`

The number fluctuate a bit, I think there's no performance gain/loss.

I was able to reproduce the memoryview regression that @tylerjereddy flagged up, but I think it's an order of magnitude smaller than the numbers measured here, so it won't show up.  It's something to keep in mind if we're writing code that has a lot of function calls (there's 3 here).

re: precision, yes I had to make the qcp function use doubles throughout to keep it working.  The `InnerProduct` can use either float or double now.  Strangely these take the same amount of time, which means the compiler isn't able to do anything clever here (where it ought to be) so probably some future digging.  This does some ground work at least.

**Review by orbeckst — APPROVED** [review]

**Pull request merged**